lib/path.nix

```
@@ -0,0 +1,172 @@
+      # matches patterns like "/", "/./", "/././", with arbitrarily many "/"s
+      # together. These are the main special cases:
+      # - Leading "./" gets split into a leading "." part
+      # - Trailing "/." or "/" get split into a trailing "." or ""
+      #   part respectively
+      #
+      # These are the only cases where "." and "" parts can occur
+      parts = split "/+(\\./+)*" path;
+
+      # `split` creates a list of 2 * k + 1 elements, containing the k +
+      # 1 parts, interleaved with k matches where k is the number of
+      # (non-overlapping) matches. This calculation here gets the number of parts
+      # back from the list length
+      # floor( (2 * k + 1) / 2 ) + 1 == floor( k + 1/2 ) + 1 == k + 1
+      partCount = length parts / 2 + 1;
+
+      # To assemble the final list of components we want to:
+      # - Skip a potential leading ".", normalising "./foo" to "foo"
+      # - Skip a potential trailing "." or "", normalising "foo/" and "foo/." to
+      #   "foo"
+      skipStart = if head parts == "." then 1 else 0;
+      skipEnd = if last parts == "." || last parts == "" then 1 else 0;
+
+      # We can now know the length of the result by removing the number of
+      # skipped parts from the total number
+      componentCount = partCount - skipEnd - skipStart;
+
+    in
+      # Special case of a single "." path component. Such a case leaves a
+      # componentCount of -1 due to the skipStart/skipEnd not verifying that
+      # they don't refer to the same character
+      if path == "." then []
+
+      # And we can use this to generate the result list directly. Doing it this
+      # way over a combination of `filter`, `init` and `tail` makes it more
+      # efficient, because we don't allocate any intermediate lists
+      else genList (index:
+        # To get to the element we need to add the number of parts we skip and
+        # multiply by two due to the interleaved layout of `parts`
+        elemAt parts ((skipStart + index) * 2)
+      ) componentCount;
+
+  # joins subpath components together
+  joinSubpath = components:
+    # Always return relative paths with `./` as a prefix (./path.md#leading-dots-for-subpaths)
+    "./" +
+    # An empty string is not a valid subpath, so we need to return a `.` when we have no components
+    (if components == [] then "." else concatStringsSep "/" components);
+
+in /* No rec! Add dependencies on this file at the top. */ {
+
+  /* Normalise a subpath.
+
+  - Limit repeating `/` to a single one
+
+  - Remove redundant `.` components
+
+  - Error on empty strings
+
+  - Remove trailing `/` and `/.`
+
+  - Error on `..` path components
+
+  - Add leading `./`
+
+  Type:
+    subpath.normalise :: String -> String
+
+  Laws:
+
+  - (Idempotency) Normalising multiple times gives the same result:
+
+        subpath.normalise (subpath.normalise p) == subpath.normalise p
+
+  - (Uniqueness) There's only a single normalisation for a path:
```

> There's only a single normalisation for a path

This reads like univalence, which holds for all (pure) functions.

```suggestion
  - (Uniqueness) There's only a single normalisation for the paths that lead to the same file system node:
```